### DISCUSSION TITLE
Genre Suggestion and Question About a User Manual

### DISCUSSION THREAD

**Discussion by BlakeMulligan** [rdswift/picard-scripts#2] [Ideas & Requests]
Your quick responses have me in appreciation. 

I'm wondering if there is a way to set Genre for artist ... example Travis Tritt ... his songs can generate a lot of "genre tags" reality is they are mostly sub categories. Country is his artist genre but every plugin that I've seen insists on using tags for individual tracks. Option 2 would be a preferred genre ... where a limited number of options. 

"if ______ is a tag option in a group, set ______." probably too complicated but I personally get annoyed to find an artist that I know fits a simple genre (rock/country/blues) has a few tracks that get sub categories so the song doesn't show in their main.

Travis Tritt - default would be country: so if tags are country rock, alt country, indie country would all get flagged and tag would set to country instead. Have to be an opt in setting of course, if it's even possible to do. 

Have I missed a user guide that tells us how to do things already available instead of bothering you with how do I comments on the forum? If there isn't want I would be willing to collaborate with you to help create one. 

By the way almost two years ago I started working on my collection but pressing health issues put it all on hold. I asked back then about sorting to Genre based directory. My understanding it wasn't quite doable based on what Picard had function wise. I did figure out a work around - still slow but can be done over time. Get everything sorted so all from an artist is together tag all so only one genre then on line 461 insert %genre% as shown here 

$rreplace(%genre%/%_nFilePath%/%_nFileName%,[?*:\\_]+,_)

it of course returns the result of Destination Directory\<genre>\renamed string

"R:\Music Check\Country\~Alpha~\<Artist>\[YEAR] <Album Title>\<File Type>\<Trac #> <Trac Title>

A little extra work but since I split my music over several drives and prefer to do so by genre this helps for the sort. Then I just grab the appropriate directory and move to the storage drive.

**Comment by rdswift** [comment]
I think setting a genre for an artist (in addition to setting genre on tracks) is not yet implemented in the MusicBrainz database.  Until it is, that information isn't available to Picard directly, so anything we do will be a work-around.  This is a bit of a hack, but you could have a number of tagging scripts (not enabled) that set a new variable such as `_my_genre` (one script per genre), like:

```taggerscript
$set(_my_genre,Country)
```

and in your file naming script wherever you want the genre to appear in the path include `$if2(%_my_genre%,Unspecified)`.  Then before saving an album in the right-hand pane, right-click on the album (not one of the tracks) and run the script that you created to set the appropriate genre.  Then when you save, it should be saved to the appropriate genre directory.  Alternately, if most of your music is one genre (or you're willing to process mostly one genre at a time) you can enable the script setting that genre as the default, and then just override it if necessary by doing the right-click and run a different genre script.

As for a manual, there is the [Picard Users Guide](https://picard-docs.musicbrainz.org) that I put together (with input from the community) and try to keep up-to-date.  There is a lot of information there, including a tutorial on developing your own file naming script.  If you haven't seen it, I encourage you to check it out.  And if you have questions or suggestions, please raise them.

**Reply by BlakeMulligan** [reply]
Thank you. I will consider that scripting option. It will make the processing easier than my current method ... right click paste in genre to the whole album / collection. Script makes so much more sense. I have a friend who I just recently shared Picard and your script with she's a classical listener and is excited by the possibilities

I will check out that guide. If I can think of anything to add to that I will let you know. Have a great night and day tomorrow.

**Reply by BlakeMulligan** [reply]
Just to be clear I was thinking it would be useful to have a guide for your script specifically. It seems there are a few options that people might want to know how to turn on or off without bothering you for help. I know the thread on your script is long and I did try ot search but didn't find any help so I had to ask. So if there is one for your script that's fantastic if now I can try to put some notes together from your long thread, get a feel for what I can then hit you up for any clarification. Just a thought. I should have some time the next few months while we deal with things here.

**Reply by rdswift** [reply]
> Just to be clear I was thinking it would be useful to have a guide for your script specifically.

Sorry, I misunderstood.  There isn't one, and I never thought about creating one mostly because it sort of evolved over time and I tried to keep it documented with comments.  I can certainly prepare something if you think there's value in it, but should it have a user focus (explaining how the various options impact the output) or a developer focus (explaining how it works)?

**EDIT:** Any thoughts or suggestions regarding layout or content are much appreciated.  I'm usually too close to the script to see things from the perspective of someone not as familiar with how it works and how to use it.  Thanks.

**Reply by rdswift** [reply]
Just thinking that perhaps the [README](https://github.com/rdswift/picard-scripts/tree/master/naming_scripts/Bob_Swift) could be expanded to include the information?

**Reply by BlakeMulligan** [reply]
I will dive into the readme. 

To answer you questions. Both areas make sense. I know I have somethings I might like to see in a script and if I understood how it works I might be able to code them myself. But the user aspect also makes sense to know where there are things that you can opt into or out of. 

One thing I would like to switch but realize it's my need and not likely one for others ... track numbering I just personally like 01, 02, 03 ... so when I hit 10, 11, etc it all lines up. Just a stupid detail I'm sure. 

Oh and I did try out the idea of the script to run the genre ... it works. a little cumbersome but in lieu of any better options it works fine. I know once I get things cleaned up that I won't care so much that I have to run it. 

I've even thought about trying to write an if statement sort but it probably isn't feasible. Where user could define as follows 

if genre = ________ (a list of input by or turned on by the user) then rewrite the genre to _________ 

I know it could never be perfect but keeping as much automated as possible would be great. I may just give that a go in a month or so when things hopefully calm down. But once again your quick and simple solution works and I've developed a system to catch my mistakes. 

I will give you more feedback soon after I really look at the readme. Confession time I didnt even read it before because most readme's have very little in them beyond disclaimers. Though it might be a product of my main work environment and the related applications where readme is mostly legal mumbo jumbo that I have to accept or the software won't work

**Reply by rdswift** [reply]
> To answer you questions. Both areas make sense. I know I have somethings I might like to see in a script and if I understood how it works I might be able to code them myself. But the user aspect also makes sense to know where there are things that you can opt into or out of.

I was afraid you'd say that (even though I was already thinking it as well).  :stuck_out_tongue_winking_eye: 

> One thing I would like to switch but realize it's my need and not likely one for others ... track numbering I just personally like 01, 02, 03 ... so when I hit 10, 11, etc it all lines up. Just a stupid detail I'm sure.

The script should already be doing that for the file naming.  The minimum padding is set in Line 82 and the required padding is calculated in Lines 248-257.  The actual padding is done in Lines 259-266.

> Oh and I did try out the idea of the script to run the genre ... it works. a little cumbersome but in lieu of any better options it works fine. I know once I get things cleaned up that I won't care so much that I have to run it.
> 
> I've even thought about trying to write an if statement sort but it probably isn't feasible. Where user could define as follows
> 
> if genre = ________ (a list of input by or turned on by the user) then rewrite the genre to _________
>

You could do that in an enabled tagging script like:

```taggerscript
$setmulti(_country,country; country & western; c&w; rodeo)
$if($inmulti($lower(%genre%),%_country),$set(genre,Country)

$setmulti(_rock,rock; progressive rock; classic rock; hard rock; metal; heavy metal)
$if($inmulti($lower(%genre%),%_rock),$set(genre,Rock)
```

I also created a plugin that allowed the user to do some of this automatically, including optionally setting up regular expressions for matching.  I never submitted it to the list of "official" plugins, but it is available at https://github.com/rdswift/picard-plugins/tree/2.0_RDS_Plugins/plugins/genre_mapper in case you find it useful.

> I will give you more feedback soon after I really look at the readme. Confession time I didnt even read it before because most readme's have very little in them beyond disclaimers. Though it might be a product of my main work environment and the related applications where readme is mostly legal mumbo jumbo that I have to accept or the software won't work

I know what you mean.  I usually try to include a bit more of a mini user guide to at least provide a starting point.

**Reply by BlakeMulligan** [reply]
Another operator error. sorry about the numbering comment it does exactly what it should ... have to not screw up and run a different renaming I was testing. DOH!

Seriously it amazes me how easy you can write solutions. I promise I will delve into the script and help write at least the user portion. If you're interested that is. What I'm thinking is I can dive in write outline / key points, where I can I will help with developer notes. We could get it flushed out a little and perhaps get a few other users to help. Of course you may have some power users / developers that could jump in and not need input from me. I won't be offended if that is the case. I just figured since I brought this up I should offer to help and would be glad to. 

I will test out the genre scripts both the lines you put (is there a place they could drop into your script or do they need to run separately) here and the script you never submitted.

**Reply by rdswift** [reply]
> Another operator error. sorry about the numbering comment it does exactly what it should ... have to not screw up and run a different renaming I was testing. DOH!

No apology necessary.  That happens to all of us.

> Seriously it amazes me how easy you can write solutions.

Not always that easy, but I have been writing code for about 50 years now.  I've only been working on Picard for about 5 years, and there's still so much I don't understand.

> I promise I will delve into the script and help write at least the user portion. If you're interested that is. What I'm thinking is I can dive in write outline / key points, where I can I will help with developer notes. We could get it flushed out a little and perhaps get a few other users to help. Of course you may have some power users / developers that could jump in and not need input from me. I won't be offended if that is the case. I just figured since I brought this up I should offer to help and would be glad to.

Any help would be much appreciated, even an outline of what you would like to see covered.  That way I have a bit of a road map to work from.

> I will test out the genre scripts both the lines you put (is there a place they could drop into your script or do they need to run separately) here and the script you never submitted.

I suggest adding it as a new tagging script (not in the file naming script) because file naming scripts don't actually update the values saved to tags in the files.  Also, I suggest enabling it so that you don't have to run it manually.

**Reply by BlakeMulligan** [reply]
Bob, 

I wanted to apologize for dropping off. Shortly after we last spoke my close friend Jacque's health took a bad turn. After a 5 years battle with cancers we lost her just days before Christmas. It's been a lot of work dealing with estate issues and only beginning to see a light at the end of the tunnel. 

If you're still interested in my input I will try to look things over and send you some notes soon. 

I do hope you are well and thanks for all you do.

Will

**Reply by rdswift** [reply]
Sorry to hear about your friend.  I know how much work it is to deal with estate issues, having gone through it a couple of times myself.

If you're still interested in sending me notes, I'll review them and see what I can do.  Thanks.
